File: 0_3.leetcode/java/Search_In_Rotated_Sorted_Array.java
public class Search_In_Rotated_Sorted_Array {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = (right + left) / 2;

            if (nums[mid] <= nums[right]) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        int pivot = right;
        left = -1;
        right = nums.length;

        while (left + 1 < right) {
            int mid = (right + left) / 2;
            int midPivot = (mid + pivot) % nums.length;

            if (nums[midPivot] < target) {
                left = mid;
            } else if (nums[midPivot] > target) {
                right = mid;
            } else {
                return midPivot;
            }
        }
        return -1;
    }
}

File: 0_3.leetcode/kotlin/src/main/kotlin/Search_In_Rotated_Sorted_Array.kt
fun main() {
    val result = search(intArrayOf(1, 3, 5), 5)
    println("result = ${result}")
}
fun search(nums: IntArray, target: Int): Int {
    var left = 0
    var right = nums.size - 1

    while (left < right) {
        val mid = (left + right) / 2
        if (nums[mid] <= nums[right]) {
            right = mid
        } else {
            left = mid + 1;
        }
    }

    val pivot = right
    left = -1
    right = nums.size

    while (left + 1 < right) {
        val mid = (left + right) / 2
        val midPivot = (mid + pivot) % nums.size

        if (nums[midPivot] < target) {
            left = mid
        } else if (nums[midPivot] > target) {
            right = mid
        } else {
            return midPivot
        }
    }
    return -1
}